components/TextFields/src/ContainedInputView/private/MDCContainedInputViewClearButton.m

```diff
@@ -0,0 +1,197 @@
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.50083 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.45367 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.31467 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.26750 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.26750 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.31467 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.45367 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.50083 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.26750 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.68700 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.31467 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.73417 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.50083 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.54800 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.68700 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.73417 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.73417 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.68700 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.54800 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.50083 * innerBounds.size.height)];
+  [ic_clear_path
+      addLineToPoint:CGPointMake(
+                         CGRectGetMinX(innerBounds) + (CGFloat)0.73417 * innerBounds.size.width,
+                         CGRectGetMinY(innerBounds) + (CGFloat)0.31467 * innerBounds.size.height)];
+  [ic_clear_path closePath];
+
+  return ic_clear_path;
+}
+
+- (CGFloat)imageViewSideLength {
+  return kClearButtonInnerImageViewSideLength;
+}
+
+- (CGFloat)sideLength {
+  return kClearButtonTouchTargetSideLength;
+}
+
+@end
```

Comment:
Based on the snapshots it looks like this PR is missing layout logic. Please add it so that the snapshot tests look as expected.

Reply:
Is it expected that the TextFields are usable with every small PR? And are you okay with a new layout object class being introduced in this PR? Will that make it too big?

Reply:
Yes, they should be functional.

Do you need a layout class object class in order to lay the clear button out?

Reply:
I didn't realize that. I want to block their use internally until they've moved over fully. I also don't really expect them to be used externally. Is this negotiable?

The reason I ask about a layout object is because eventually the clear button will get its frame from the layout object. I'm trying to avoid writing a bunch of new code that I plan to just get rid of a few PRs down the road. Does that make sense?

Reply:
Because of the complexity of this component, it's important that we approach its construction through an iterative evolution of functional products. There's a classic diagram related to this mindset:

![1_-Uw18kKkuDk6mK01gg0zUg](https://user-images.githubusercontent.com/45670/62731783-7d8bc380-ba12-11e9-8db2-9f858f23dd1d.png)

My intended meaning of sharing this chart is that each step of the implementation of a product should follow a **logical and functional** iteration. Does this mean we might do some extra work re-creating things? Yes. But the end result will be better because of it.

Your mind contains more knowledge today than it did eight months ago, full stop. Use that additional knowledge to approach building this component in the best way you can today, instead of the best way you could 8 months ago.

> The reason I ask about a layout object is because eventually the clear button will get its frame from the layout object.

I don't see a strong argument for having a layout object yet. Once there's a clear case to be made for that, we can refactor the relevant parts where they need to go and improve the code at that point.

Reply:
Does you not seeing a strong argument for having a layout object at this juncture mean that if I include one you will veto it?

Reply:
I think it's probably overkill for this particular problem and would like to better understand  how, when, and why a layout object becomes an important part of this implementation.

Reply:
The layout object I would bring in right now would not do most of the things its more full fledged version in the experimental directory does. It would be a good place to calculate where this custom clear button would go in isolation from other concerns of the view. I tried assigning the clear button frame to `-clearButtonRectForBounds:` on super and it didn't look good :/

Maybe the clear button wasn't the best place to start. What if I opened another PR that introduces a layout object that concerns itself only with positioning the leading/trailing views in a way that's more in accordance with the guidelines than what's currently there?

Reply:
It feels like you're working backwards from an assumed solution, rather than forwards from where the code is right now. It's reasonable for us to add a clear button at this juncture in the code, but I'm not 100% convinced that a layout object is necessary to do so. That I'm not convinced right now doesn't mean I can't be convinced, though, so pushing up the code for the layout is the best way to start having a concrete discussion.

Reply:
I think it makes more sense for the first thing the layout object does to be the leading/trailing view calculation, so I'm going to open a PR that does that, then I'll come back to this one.